flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/hybrid/HybridSourceITCase.java

```diff
@@ -0,0 +1,266 @@
+                            int numSplits = 2;
+                            int numRecordsPerSplit = 10;
+                            int startingValue = 20;
+                            Boundedness boundedness = Boundedness.BOUNDED;
+                            // TODO: convert source parameters to splits from
+                            // MockBaseSource.createEnumerator - this should be a
+                            // shared utility
+                            List<MockSourceSplit> splits = new ArrayList<>();
+                            for (int i = 0; i < numSplits; i++) {
+                                int endIndex =
+                                        boundedness == Boundedness.BOUNDED
+                                                ? numRecordsPerSplit
+                                                : Integer.MAX_VALUE;
+                                MockSourceSplit split = new MockSourceSplit(i, 0, endIndex);
+                                for (int j = 0; j < numRecordsPerSplit; j++) {
+                                    split.addRecord(startingValue + i * numRecordsPerSplit + j);
+                                }
+                                splits.add(split);
+                            }
+                            return splits;
+                        }));
+        return new HybridSource<>(sourceChain);
+    }
+
+    private void testHybridSource(FailoverType failoverType, Source source) throws Exception {
+
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(PARALLELISM);
+        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(1, 0));
+
+        final DataStream<Integer> stream =
+                env.fromSource(source, WatermarkStrategy.noWatermarks(), "hybrid-source")
+                        .returns(Integer.class);
+
+        final DataStream<Integer> streamFailingInTheMiddleOfReading =
+                RecordCounterToFail.wrapWithFailureAfter(stream, EXPECTED_RESULT.size() / 2);
+
+        final ClientAndIterator<Integer> client =
+                DataStreamUtils.collectWithClient(
+                        streamFailingInTheMiddleOfReading,
+                        HybridSourceITCase.class.getSimpleName() + '-' + failoverType.name());
+        final JobID jobId = client.client.getJobID();
+
+        RecordCounterToFail.waitToFail();
+        triggerFailover(
+                failoverType,
+                jobId,
+                RecordCounterToFail::continueProcessing,
+                miniClusterResource.getMiniCluster());
+
+        final List<Integer> result = new ArrayList<>();
+        while (result.size() < EXPECTED_RESULT.size() && client.iterator.hasNext()) {
+            result.add(client.iterator.next());
+        }
+
+        verifyResult(result);
+    }
+
+    // ------------------------------------------------------------------------
+    //  test utilities
```

Comment:
should we extract those source test utils into `flink-connector-test-utils` to avoid code duplication?